doc/ADR/cpp_api_modularisation.md

```diff
@@ -0,0 +1,103 @@
+\page cpp-api-and-modularisation Libcprover-cpp and Modularisation
+
+**Date**: 22 Jun 2023
+**Updated**: 22 Jun 2023
+**Author**: Fotis Koutoulakis, fotis.koutoulakis@diffblue.com
+**Domain**: Architecture, API
+**Description**: This document outlines our thinking about the rearchitecting of
+    CBMC using the C++ API (`Libcprover-cpp`) as the central module and the
+    transitioning of other tools to use that as a basis.
+
+## Motivation && Current State of Affairs
+
+CProver is a collection of tools fitting various program analysis needs. It
+has been the product of the evolution of the codebase of the model-checking
+tool for C (`CBMC`), after it has been adopted with various front-ends/back-ends
+and various auxilliary tools.
+
+During this time, the repository has grown organically, using some guidelines
+for development that were based on tradition and intuition rather than some
+tight development guide. This development model has been successful for most
+of CProver's life, based on its nature as a hybrid industrial/academic and
+experimental/applied tool. However, it has had the side-effect of accruing
+some code duplication and technical debt. It has also made the codebase harder
+to understand/develop and integrate with, especially for someone who is new
+to the codebase.
+
+The above concerns have generated discussions about the breaking down of
+CProver into modules, with cleaner interfaces and tighter boundary control
+between them, making the code easier to integrate in other projects (by making
+the various component modules easier to combine and reuse) and making the
+codebase easier to understand and maintain.
+
+## The Plan Going Forward
+
+Given the above outline, we have reached a point where we are strongly motivated
+to take action to better componentise and modularise CProver.
+
+It's also an opportune time for us, given the existence of `Libcprover-cpp`
+the C++ API that we built to support interfacing with Rust (for now - other
+languages may be coming in the future): we can use this as the basis of
+development for an API exposing the interfaces of the various other modules
+and refactor them into the better-defined shape we want them to take on
+an incremental basis.
+
+Of course, this is a project that is massive in scope, potentially being
+exposed to further scope creep. We acknowledge that any effort to do what
+we have discussed already is going to be a multi-year effort from our end,
+and that we will need community alignment to achieve the outcome we're
+looking for.
+
+This is why we are looking into testing the approach on a smaller component
+first, to get a better feel for the amount of effort and any challenges
+lurking in the dark.
+
+## `goto-bmc` and `libcprover-cpp`
```

Comment:
There is a fairly straight forward implication in the architecture here, which is not yet explicitly discussed in this document. This is that if `goto-bmc` is written using the `libcprover-cpp` interface only, then as a necessity it will not be following precisely the same code-patterns as the other command line tools due to the differing implementation choice. Should we include a statement of intent in this document for the future of the existing user interfaces, in order to achieve consistency in the long term?

Reply:
Yeah I added a paragraph making this explicit.